Sources/TuistKit/Utils/RootDirectoryLocator.swift

```diff
@@ -15,6 +15,12 @@ final class RootDirectoryLocator: RootDirectoryLocating {
     /// This cache avoids having to traverse the directories hierarchy every time the locate method is called.
     fileprivate var cache: [AbsolutePath: AbsolutePath] = [:]
 
+    /// Shared instance
+    static var shared: RootDirectoryLocating = RootDirectoryLocator()
```

Comment:
> [minor/personal opinion here - please take my comments with a pinch of salt 🧂]

Thanks for keep sharing those because it's always better to have always different points of view.

I think dependency injection is another valid approach, but I had to make a trade-off here _(like most of the times in programming)_. The risk of having shared instances that hold state is that they might have side effects that might get out of control. I think it's not the case of `RootDirectoryLocator` where the state is private and changes in the state are not relevant to the rest of the code base.

Dependency injection comes with its downsides too. Swift doesn't make things easier and that results in some boilerplate code that is hard to reason about. In my experience, I always find very hard o reason about injected dependencies _(unless they were injected through the constructor)_. When we build abstractions for easing dependency injection, or taking it to a whole new level, reasoning about the code becomes really hard: *what am I getting here?* *who is injecting the dependency?*, *from where is the dependency injected?*

I'm happy to reconsider the approach if we see the current one resulting in side effects that make the code execution go through unexpected paths. Given the amount of global dependencies that we have at the moment, and the fact that they are not seeing big issues with it, I'd not focus on this at the moment.